external-crates/move/documentation/book/src/control-flow/pattern-matching.md

```diff
@@ -0,0 +1,618 @@
+match on an expression which differs from the type of the pattern in the match this will result in a
+type error. For example:
+
+```move
+match (1) {
+    // The `true` literal pattern is of type `bool` so this is a type error.
+    true => 1,
+    // TYPE ERROR: expected type u64, found bool
+    _ => 2,
+}
+```
+
+Similarly the following would also result in a type error since `MyEnum` and `MyStruct` are
+different types:
+
+```
+match (MyStruct { x: 0, y: 0 }) {
+    MyEnum::Variant(..) => 1,
+    // TYPE ERROR: expected type MyEnum, found MyStruct
+}
+```
+
+## Matching
+
+Prior to delving into the specifics of pattern matching and what it means for a value to "match" a
+pattern, let's examine a few examples to provide an intuition for the concept.
+
+```move
+fun test_lit(x: u64): u8 {
+    match (x) {
+        1 => 2,
+        2 => 3,
+        _ => 4,
+    }
+}
+test_lit(1); // returns 2
+test_lit(2); // returns 3
+test_lit(3); // returns 4
+test_lit(10); // returns 4
+
+fun test_var(x: u64): u64 {
+    match (x) {
+        y => y,
+    }
+}
+test_var(1); // returns 1
+test_var(2); // returns 2
+test_var(3); // returns 3
+...
+
+const MyConstant: u64 = 10;
+fun test_constant(x: u64): u64 {
+    match (x) {
+        MyConstant => 1,
+        _ => 2,
+    }
+}
+test_constant(MyConstant); // returns 1
+test_constant(10); // returns 1
+test_constant(20); // returns 2
+
+fun test_or_pattern(x: u64): u64 {
+    match (x) {
+        1 | 2 | 3 => 1,
+        4 | 5 | 6 => 2,
+        _ => 3,
+    }
+}
+test_or_pattern(3); // returns 1
+test_or_pattern(5); // returns 2
+test_or_pattern(70); // returns 3
+
+fun test_or_at_pattern(x: u64): u64 {
+    match (x) {
+        x @ (1 | 2 | 3) => x + 1,
+        y @ (4 | 5 | 6) => y + 2,
+        z => z + 3,
+    }
+}
+test_or_pattern(2); // returns 3
+test_or_pattern(5); // returns 7
+test_or_pattern(70); // returns 73
+```
+
+The most important thing to note from these examples is that a pattern matches a value if the value
+is equal to the pattern, and wildcard/variable patterns match anything. This is true for literals,
+variables, and constants. For example, in the `test_lit` function, the value `1` matches the pattern
+`1`, the value `2` matches the pattern `2`, and the value `3` matches the wildcard `_`. Similarly,
+in the `test_var` function, the value `1` matches the pattern `y` and the value `2` matches the
+pattern `y`.
+
+A variable `x` matches (or "equals") any value, and a wildcard `_` matches any value (but only one
+value!). Or-patterns are like a logical OR, where a value matches the pattern if it matches any of
+patterns in the or-pattern so `p1 | p2 | p3` should be read "matches p1, or p2, or p3".
+
+### Matching Constructors
+
+The most interesting part of pattern matching are constructor patterns. These patterns allow you
+inspect and access deep within both structs and enums, and are the most powerful part of pattern
```

Comment:
```suggestion
inspect and access deep within both structs and enums, and are one of the most powerful parts of pattern
```
nit: unless this is quantifiable, you run the risk of insulting the other parts of pattern matching.

Reply:
Indeed, I've gotta watch out -- patterns are known to be quite vindictive when insulted like this! 

Thanks for all the suggestions!